File: ios/App/App/AppPermissionsPlugin.swift
import Foundation
import Capacitor
import UserNotifications
import Contacts
import AVFoundation
import UIKit

/**
 * JS-facing bridge for the "some permissions are missing" reminder (see
 * src/components/AppPermissionsReminder.tsx / src/lib/appPermissions.ts). Reports current grant
 * status for the handful of permissions this app actually asks for, and opens Settings to fix
 * them. There's no iOS equivalent of Android's exact-alarm-scheduling permission or battery-
 * optimization exemption, so those two just fall back to the general app Settings page too — a
 * shared JS call site still does something sensible on both platforms without needing a branch.
 */
@objc(AppPermissionsPlugin)
public class AppPermissionsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppPermissionsPlugin"
    public let jsName = "AppPermissions"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkAll", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openAppSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openExactAlarmSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openBatteryOptimizationSettings", returnType: CAPPluginReturnPromise)
    ]

    @objc func checkAll(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let notificationsGranted = settings.authorizationStatus == .authorized || settings.authorizationStatus == .provisional
            let contactsGranted = CNContactStore.authorizationStatus(for: .contacts) == .authorized
            let microphoneGranted = AVAudioSession.sharedInstance().recordPermission == .granted

            call.resolve([
                "notifications": notificationsGranted,
                "contacts": contactsGranted,
                "microphone": microphoneGranted
                // exactAlarm / batteryOptimization: Android-only concepts, deliberately omitted —
                // the JS side treats a missing key as "nothing to fix" on this platform.
            ])
        }
    }

    @objc func openAppSettings(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if let url = URL(string: UIApplication.openSettingsURLString) {
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
            call.resolve()
        }
    }

    @objc func openExactAlarmSettings(_ call: CAPPluginCall) {
        openAppSettings(call)
    }

    @objc func openBatteryOptimizationSettings(_ call: CAPPluginCall) {
        openAppSettings(call)
    }
}

File: android/app/src/main/java/com/familyledger/app/AppPermissionsPlugin.java
package com.familyledger.app;

import android.Manifest;
import android.app.AlarmManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.PowerManager;
import android.provider.Settings;
import androidx.core.content.ContextCompat;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * JS-facing bridge for the "some permissions are missing" reminder (see
 * src/components/AppPermissionsReminder.tsx). Reports current grant status for the handful of
 * permissions this app actually asks for, and opens the right OS settings screen to fix each one.
 * There's no single Capacitor-core API covering all of these together — notifications, contacts,
 * and microphone are ordinary runtime permissions, while exact-alarm scheduling and the battery-
 * optimization exemption are special OS-level toggles that live on their own dedicated Settings
 * screens, reachable only via specific Intent actions, not a permission request dialog.
 */
@CapacitorPlugin(name = "AppPermissions")
public class AppPermissionsPlugin extends Plugin {

    @PluginMethod
    public void checkAll(PluginCall call) {
        Context ctx = getContext();
        JSObject result = new JSObject();
        result.put("notifications", granted(ctx, Manifest.permission.POST_NOTIFICATIONS, Build.VERSION_CODES.TIRAMISU));
        result.put("contacts", granted(ctx, Manifest.permission.READ_CONTACTS, 0));
        result.put("microphone", granted(ctx, Manifest.permission.RECORD_AUDIO, 0));
        result.put("exactAlarm", canScheduleExactAlarms(ctx));
        result.put("batteryOptimization", isIgnoringBatteryOptimizations(ctx));
        call.resolve(result);
    }

    // `minSdk` = the API level this permission first became a runtime permission at all; below
    // that, Android grants it automatically at install time and there's nothing to check.
    private boolean granted(Context ctx, String permission, int minSdk) {
        if (minSdk > 0 && Build.VERSION.SDK_INT < minSdk) return true;
        return ContextCompat.checkSelfPermission(ctx, permission) == PackageManager.PERMISSION_GRANTED;
    }

    private boolean canScheduleExactAlarms(Context ctx) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S) return true; // no such restriction pre-Android 12
        AlarmManager am = (AlarmManager) ctx.getSystemService(Context.ALARM_SERVICE);
        return am != null && am.canScheduleExactAlarms();
    }

    private boolean isIgnoringBatteryOptimizations(Context ctx) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) return true;
        PowerManager pm = (PowerManager) ctx.getSystemService(Context.POWER_SERVICE);
        return pm != null && pm.isIgnoringBatteryOptimizations(ctx.getPackageName());
    }

    @PluginMethod
    public void openAppSettings(PluginCall call) {
        Context ctx = getContext();
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        intent.setData(Uri.parse("package:" + ctx.getPackageName()));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        ctx.startActivity(intent);
        call.resolve();
    }

    @PluginMethod
    public void openExactAlarmSettings(PluginCall call) {
        Context ctx = getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            Intent intent = new Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM);
            intent.setData(Uri.parse("package:" + ctx.getPackageName()));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            try {
                ctx.startActivity(intent);
                call.resolve();
                return;
            } catch (Exception e) {
                // Some OEM builds don't expose this screen despite advertising the API — fall
                // back to the general app settings page rather than failing the request.
            }
        }
        openAppSettings(call);
    }

    @PluginMethod
    public void openBatteryOptimizationSettings(PluginCall call) {
        Context ctx = getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
            intent.setData(Uri.parse("package:" + ctx.getPackageName()));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            try {
                ctx.startActivity(intent);
                call.resolve();
                return;
            } catch (Exception e) {
                // Same OEM-quirk fallback as above.
            }
        }
        openAppSettings(call);
    }
}
